xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,220 @@ impl CoiSystem {
+        candidates: &[String],
+        // TODO: make SMBert available to CoiSystem and remove this argument
+        smbert: F,
+    ) {
+        /// Filters the unique candidates wrt the existing key phrases.
+        fn unique_candidates<CP, F>(
+            coi: &CP,
+            candidates: &[String],
+            smbert: F,
+        ) -> BTreeSet<KeyPhrase>
+        where
+            CP: CoiPoint + CoiPointKeyPhrases,
+            F: Fn(&str) -> Result<Embedding, Error>,
+        {
+            candidates
+                .iter()
+                .filter_map(|words| {
+                    (!coi.key_phrases().contains(words))
+                        .then(|| {
+                            smbert(words)
+                                .ok()
+                                .and_then(|point| KeyPhrase::new(words, point).ok())
+                        })
+                        .flatten()
+                })
+                .collect()
+        }
+
+        /// Reduces the matrix along the axis while skipping the diagonal elements.
+        fn reduce_without_diag<S, F, G>(
+            a: ArrayBase<S, Ix2>,
+            axis: Axis,
+            reduce: F,
+            finalize: G,
+        ) -> Array2<f32>
+        where
+            S: Data<Elem = f32>,
+            F: Fn(f32, f32) -> f32 + Copy,
+            G: Fn(f32) -> f32 + Copy,
+        {
+            a.lanes(axis)
+                .into_iter()
+                .enumerate()
+                .map(|(i, lane)| {
+                    lane.iter()
+                        .enumerate()
+                        .filter_map(|(j, element)| (i != j).then(|| *element))
+                        .reduce(reduce)
+                        .map(finalize)
+                        .unwrap_or_default()
+                })
+                .collect::<Array1<_>>()
+                .insert_axis(axis)
+        }
+
+        /// Gets the index of the maximum element.
+        fn argmax<I, F>(iter: I) -> Ix
+        where
+            I: IntoIterator<Item = F>,
+            F: Borrow<f32>,
+        {
+            iter.into_iter()
+                .enumerate()
+                .reduce(|(arg, max), (index, element)| {
+                    if element.borrow() > max.borrow() {
+                        (index, element)
+                    } else {
+                        (arg, max)
+                    }
+                })
+                .map(|(arg, _)| arg)
+                .unwrap(/* at least one key phrase exists */)
+        }
+
+        /// Computes the pairwise similarity and their normalizations of the key phrases.
+        ///
+        /// The matrices are of shape `(key_phrases_len + candidates_len, key_phrases_len +
+        /// candidates_len + 1)` with the following blockwise layout:
+        /// ```text
+        /// [[sim(kp, kp),   sim(kp, cand),   sim(kp, coi)  ],
+        ///  [sim(cand, kp), sim(cand, cand), sim(cand, coi)]]
+        /// ```
+        fn similarities<CP>(
```

Comment:
Could be worth it to move this functions outside so we can unit test each one?

Reply:
as discussed, will be done in a follow up to #355